File: rasiin_design/rasiin_design/report/daily_sale/daily_sale.js
// daily_sale.js
// Copyright (c) 2025, Rasiin and contributors
// For license information, please see license.txt

frappe.query_reports["Daily Sale"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 1,
            "width": "80px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_end(),
            "reqd": 1,
            "width": "80px"
        },
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": "80px"
        },
        {
            "fieldname": "sales_person",
            "label": __("Sales Person"),
            "fieldtype": "Link",
            "options": "Sales Person",
            "width": "80px"
        },
        {
            "fieldname": "show_outstanding",
            "label": __("Show Outstanding"),
            "fieldtype": "Check",
            "default": 1,
            "width": "80px"
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (!data) return value;
		
		// Color coding for amounts
		if (column.fieldname === "total_amount" && data.total_amount > 10000) {
			value = "<span style='color: green; font-weight: bold;'>" + value + "</span>";
		}
		
		// Color coding for discount
		if (column.fieldname === "discount_amount" && data.discount_amount > 0) {
			value = "<span style='color: #ff5858;'>" + value + "</span>";
		}
		
		// Color coding for paid amount
		if (column.fieldname === "paid_amount") {
			if (data.paid_amount > 0) {
				value = "<span style='color: #28a745; font-weight: bold;'>" + value + "</span>";
			} else {
				value = "<span style='color: #6c757d;'>" + value + "</span>";
			}
		}
		
		// Color coding for outstanding amount
		if (column.fieldname === "outstanding_amount") {
			if (data.outstanding_amount > 0) {
				value = "<span style='color: #dc3545; font-weight: bold;'>" + value + "</span>";
			} else if (data.outstanding_amount < 0) {
				value = "<span style='color: #28a745; font-weight: bold;'>" + value + "</span>";
			} else {
				value = "<span style='color: #6c757d;'>" + value + "</span>";
			}
		}
		
		return value;
	},

    "onload": function(report) {
        // Initialize chart state
        report.chart_rendered = false;
        report.chart_visible = false;
        
        // Add chart toggle button
        report.page.add_inner_button(__("Toggle Charts"), function() {
            toggle_charts(report);
        });
        
        // Add refresh charts button
        report.page.add_inner_button(__("Refresh Charts"), function() {
            refresh_charts(report);
        });
        
        // Add outstanding summary button
        report.page.add_inner_button(__("Outstanding Summary"), function() {
            show_outstanding_summary(report);
        });
    },

    "after_datatable_render": function(datatable) {
        // Add custom styling
        $(datatable.wrapper).find(".dt-cell__content").css("padding", "4px 8px");
    }
};

// --- CHARTING FUNCTIONS ---

function toggle_charts(report) {
    let $chart_container = report.page.body.find(".chart-container");
    
    if ($chart_container.length === 0) {
        // Charts don't exist, render them
        render_charts(report);
        report.chart_visible = true;
    } else if ($chart_container.is(":visible")) {
        // Charts exist and are visible, hide them
        $chart_container.hide();
        report.chart_visible = false;
        frappe.show_alert({ message: __("Charts hidden"), indicator: "gray" });
    } else {
        // Charts exist but are hidden, show them
        $chart_container.show();
        report.chart_visible = true;
        frappe.show_alert({ message: __("Charts shown"), indicator: "green" });
    }
}

function refresh_charts(report) {
    // Remove existing charts
    report.page.body.find(".chart-container").remove();
    report.chart_rendered = false;
    
    // Render fresh charts
    render_charts(report);
    frappe.show_alert({ message: __("Charts refreshed"), indicator: "blue" });
}

function render_charts(report) {
    if (report.chart_rendered) return;
    
    let filters = report.get_values();

    // Create a proper container for charts
    let $chart_container = $(`
        <div class="chart-container" style="margin: 20px 0; padding: 20px; background: #f9f9f9; border-radius: 8px; border: 1px solid #d1d8dd;">
            <div style="display: flex; flex-wrap: wrap; gap: 30px; justify-content: center;">
                <div id="daily-trend-chart" style="flex: 1; min-width: 500px; min-height: 300px;"></div>
                <div id="customer-chart" style="flex: 1; min-width: 500px; min-height: 300px;"></div>
            </div>
        </div>
    `).appendTo(report.page.body);
    
    // Show loading state
    frappe.dom.freeze(__("Loading Charts..."));
    
    report.chart_rendered = true;

    frappe.call({
        method: "rasiin_design.rasiin_design.report.daily_sale.daily_sale.get_chart_data",
        args: {
            filters: filters
        },
        callback: function(r) {
            frappe.dom.unfreeze();
            
            if (r.message) {
                let chart_data = r.message;
                
                // Check if we have data for charts
                let has_daily_data = chart_data.daily_trend && chart_data.daily_trend.length > 0;
                let has_customer_data = chart_data.top_customers && chart_data.top_customers.length > 0;
                let has_outstanding_data = chart_data.outstanding_summary && chart_data.outstanding_summary.length > 0;
                
                if (has_daily_data) {
                    render_daily_trend_chart(chart_data.daily_trend);
                } else {
                    $("#daily-trend-chart").html(`
                        <div class="text-muted" style="text-align: center; padding: 50px;">
                            <i class="fa fa-bar-chart fa-2x" style="margin-bottom: 10px;"></i><br>
                            No daily sales data available
                        </div>
                    `);
                }
                
                if (has_customer_data) {
                    render_customer_chart(chart_data.top_customers);
                } else {
                    $("#customer-chart").html(`
                        <div class="text-muted" style="text-align: center; padding: 50px;">
                            <i class="fa fa-users fa-2x" style="margin-bottom: 10px;"></i><br>
                            No customer data available
                        </div>
                    `);
                }
                
                // Show message if no data at all
                if (!has_daily_data && !has_customer_data && !has_outstanding_data) {
                    $chart_container.html(`
                        <div class="text-muted" style="text-align: center; padding: 50px;">
                            <i class="fa fa-exclamation-triangle fa-2x" style="margin-bottom: 10px;"></i><br>
                            No chart data available for the selected period and filters.
                        </div>
                    `);
                }
            } else {
                // Handle API error
                $chart_container.html(`
                    <div class="text-danger" style="text-align: center; padding: 50px;">
                        <i class="fa fa-exclamation-circle fa-2x" style="margin-bottom: 10px;"></i><br>
                        Error loading chart data
                    </div>
                `);
            }
        },
        error: function() {
            frappe.dom.unfreeze();
            frappe.msgprint(__("Error loading chart data. Please try again."));
        }
    });
}

function render_daily_trend_chart(daily_data) {
    let chart = new frappe.Chart("#daily-trend-chart", {
        title: "Daily Sales Trend",
        data: {
            labels: daily_data.map(d => frappe.datetime.str_to_user(d.date)),
            datasets: [
                {
                    name: "Sales Amount",
                    values: daily_data.map(d => d.amount),
                    chartType: 'line'
                },
                {
                    name: "Discount", 
                    values: daily_data.map(d => d.discount),
                    chartType: 'line'
                }
            ]
        },
        type: 'axis-mixed',
        height: 280,
        colors: ["#5e64ff", "#ff5858"],
        axisOptions: {
            xIsSeries: true
        },
        lineOptions: {
            hideDots: 0,
            regionFill: 0
        }
    });
}

function render_customer_chart(customer_data) {
    let chart = new frappe.Chart("#customer-chart", {
        title: "Top Customers by Sales",
        data: {
            labels: customer_data.map(d => {
                // Truncate long customer names for better display
                let name = d.customer_name || d.customer || 'Unknown';
                return name.length > 20 ? name.substring(0, 20) + '...' : name;
            }),
            datasets: [{
                name: "Total Sales", 
                values: customer_data.map(d => d.total_sales)
            }]
        },
        type: 'bar',
        height: 280,
        colors: ["#743ee2"],
        barOptions: {
            spaceRatio: 0.5
        }
    });
}

function show_outstanding_summary(report) {
    let filters = report.get_values();
    
    frappe.call({
        method: "rasiin_design.rasiin_design.report.daily_sale.daily_sale.get_outstanding_summary",
        args: {
            filters: filters
        },
        callback: function(r) {
            if (r.message) {
                let summary_data = r.message;
                show_outstanding_dialog(summary_data);
            }
        }
    });
}

function show_outstanding_dialog(summary_data) {
    let dialog = new frappe.ui.Dialog({
        title: __('Outstanding Amount Summary'),
        width: 800
    });
    
    let content = '';
    
    // Total outstanding summary
    content += '<div class="alert alert-info" style="margin-bottom: 20px;">';
    content += '<h5>' + __('Total Outstanding: ') + '<strong>' + safe_format_currency(summary_data.total_outstanding || 0) + '</strong></h5>';
    content += '<p class="small text-muted mb-0">' + __('As of selected period') + '</p>';
    content += '</div>';
    
    // Outstanding by customer
    content += '<h5>' + __('Outstanding by Customer') + '</h5>';
    content += '<div class="table-responsive" style="max-height: 400px; overflow-y: auto;">';
    content += '<table class="table table-bordered table-sm">';
    content += '<thead><tr>' +
        '<th>' + __('Customer') + '</th>' +
        '<th>' + __('Total Sales') + '</th>' +
        '<th>' + __('Total Paid') + '</th>' +
        '<th>' + __('Outstanding') + '</th>' +
        '</tr></thead><tbody>';
    
    if (summary_data.customer_outstanding && summary_data.customer_outstanding.length > 0) {
        summary_data.customer_outstanding.forEach(function(row) {
            let outstanding_class = row.outstanding_amount > 0 ? 'text-danger' : (row.outstanding_amount < 0 ? 'text-success' : 'text-muted');
            content += '<tr>' +
                '<td>' + (row.customer_name || row.customer || __('Unknown')) + '</td>' +
                '<td class="text-right">' + safe_format_currency(row.total_sales || 0) + '</td>' +
                '<td class="text-right">' + safe_format_currency(row.total_paid || 0) + '</td>' +
                '<td class="text-right ' + outstanding_class + '"><strong>' + safe_format_currency(row.outstanding_amount || 0) + '</strong></td>' +
                '</tr>';
        });
    } else {
        content += '<tr><td colspan="4" class="text-center text-muted">' + __('No outstanding data found') + '</td></tr>';
    }
    content += '</tbody></table></div>';
    
    // Aging summary if available
    if (summary_data.aging_summary) {
        content += '<h5 class="mt-4">' + __('Aging Summary') + '</h5>';
        content += '<div class="table-responsive">';
        content += '<table class="table table-bordered table-sm">';
        content += '<thead><tr>' +
            '<th>' + __('0-30 Days') + '</th>' +
            '<th>' + __('31-60 Days') + '</th>' +
            '<th>' + __('61-90 Days') + '</th>' +
            '<th>' + __('Over 90 Days') + '</th>' +
            '</tr></thead><tbody><tr>';
        
        content += '<td class="text-right ' + (summary_data.aging_summary.range1 > 0 ? 'text-warning' : 'text-muted') + '">' + 
            safe_format_currency(summary_data.aging_summary.range1 || 0) + '</td>';
        content += '<td class="text-right ' + (summary_data.aging_summary.range2 > 0 ? 'text-warning' : 'text-muted') + '">' + 
            safe_format_currency(summary_data.aging_summary.range2 || 0) + '</td>';
        content += '<td class="text-right ' + (summary_data.aging_summary.range3 > 0 ? 'text-danger' : 'text-muted') + '">' + 
            safe_format_currency(summary_data.aging_summary.range3 || 0) + '</td>';
        content += '<td class="text-right ' + (summary_data.aging_summary.range4 > 0 ? 'text-danger' : 'text-muted') + '">' + 
            safe_format_currency(summary_data.aging_summary.range4 || 0) + '</td>';
        content += '</tr></tbody></table></div>';
    }
    
    dialog.$body.html(content);
    dialog.show();
}

// Safe currency formatting without recursion
function safe_format_currency(amount) {
    if (amount === null || amount === undefined) return '0.00';
    
    // Simple formatting without using frappe.format to avoid recursion
    var formatted = parseFloat(amount).toFixed(2);
    var parts = formatted.split('.');
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    
    // Get currency symbol from system defaults
    var currency_symbol = frappe.boot.sysdefaults.currency_symbol || '$';
    
    return currency_symbol + ' ' + parts.join('.');
}

File: rasiin_design/rasiin_design/report/daily_sale/daily_sale.py
# daily_sale.py
# Copyright (c) 2025, Rasiin and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, nowdate, flt, cint
import json
from datetime import datetime, timedelta

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_columns(filters)
    data = get_data(filters)
    
    return columns, data
def get_columns(filters):
    """Define report columns"""
    columns = [
        {
            "fieldname": "posting_date",
            "label": _("Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "customer",
            "label": _("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": 120
        },
        {
            "fieldname": "customer_name",
            "label": _("Customer Name"),
            "fieldtype": "Data",
            "width": 180
        },
        {
            "fieldname": "voucher_type",
            "label": _("Voucher Type"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "voucher_no",
            "label": _("Voucher No"),
            "fieldtype": "Dynamic Link",
            "options": "voucher_type",
            "width": 140
        },
        {
            "fieldname": "discount_amount",
            "label": _("Discount Amount"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "total_amount",
            "label": _("Total Amount"),
            "fieldtype": "Currency",
            "width": 120
        }
    ]
    
    # Add paid amount column
    columns.append({
        "fieldname": "paid_amount",
        "label": _("Paid Amount"),
        "fieldtype": "Currency",
        "width": 120
    })
    
    # Add outstanding amount column if enabled
    if filters.get("show_outstanding"):
        columns.append({
            "fieldname": "outstanding_amount",
            "label": _("Outstanding Amount"),
            "fieldtype": "Currency",
            "width": 130
        })
    
    columns.extend([
        {
            "fieldname": "payment_mode",
            "label": _("Payment Mode"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "custom_refrence",
            "label": _("Reference"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "sales_person",
            "label": _("Sales Person"),
            "fieldtype": "Data",
            "width": 120
        }
    ])
    
    return columns

def get_data(filters):
    """Fetch and process sales data based on filters"""
    
    conditions, condition_values = get_conditions(filters)
    
    # Main query to get sales data grouped by invoice
    outstanding_field = ", si.outstanding_amount" if filters.get("show_outstanding") else ""
    
    query = """
        SELECT 
            si.posting_date,
            si.customer,
            si.customer_name,
            'Sales Invoice' as voucher_type,
            si.name as voucher_no,
            si.discount_amount,
            si.grand_total as total_amount,
            (si.grand_total - si.outstanding_amount) as paid_amount
            {outstanding_field},
            (SELECT GROUP_CONCAT(sip.mode_of_payment SEPARATOR ', ') 
             FROM `tabSales Invoice Payment` sip 
             WHERE sip.parent = si.name) as payment_mode,
             si.custom_refrence as custom_refrence,
            (SELECT GROUP_CONCAT(DISTINCT sp.sales_person SEPARATOR ', ') 
             FROM `tabSales Team` sp 
             WHERE sp.parent = si.name) as sales_person
        FROM `tabSales Invoice` si
        WHERE si.docstatus = 1 {conditions}
        ORDER BY si.posting_date DESC, si.name
    """.format(
        outstanding_field=outstanding_field,
        conditions=conditions
    )
    
    try:
        data = frappe.db.sql(query, condition_values, as_dict=1)
    except Exception as e:
        frappe.throw(_("Error fetching data: {0}").format(str(e)))
    
    # Process and enhance data
    processed_data = process_data(data, filters)
    
    return processed_data

def get_conditions(filters):
    """Build conditions based on filters"""
    conditions = []
    condition_values = {}
    
    if filters.get("from_date"):
        conditions.append("si.posting_date >= %(from_date)s")
        condition_values["from_date"] = filters.get("from_date")
    if filters.get("to_date"):
        conditions.append("si.posting_date <= %(to_date)s")
        condition_values["to_date"] = filters.get("to_date")
    if filters.get("customer"):
        conditions.append("si.customer = %(customer)s")
        condition_values["customer"] = filters.get("customer")
    if filters.get("sales_person"):
        conditions.append("si.name IN (SELECT parent FROM `tabSales Team` WHERE sales_person = %(sales_person)s)")
        condition_values["sales_person"] = filters.get("sales_person")
    
    # FIX: Properly handle empty conditions
    if conditions:
        conditions_str = " AND " + " AND ".join(conditions)
    else:
        conditions_str = ""
    
    return conditions_str, condition_values

def process_data(data, filters):
    """Process and enhance the raw data"""
    processed_data = []
    
    for row in data:
        # Ensure numeric values
        row.discount_amount = flt(row.discount_amount)
        row.total_amount = flt(row.total_amount)
        row.paid_amount = flt(row.paid_amount)
        
        # Calculate outstanding amount if not already included
        if filters.get("show_outstanding") and 'outstanding_amount' not in row:
            row.outstanding_amount = get_outstanding_amount(row.voucher_no)
        elif filters.get("show_outstanding"):
            row.outstanding_amount = flt(row.outstanding_amount)
        
        # Format payment mode and sales person for better display
        if row.payment_mode:
            row.payment_mode = row.payment_mode.replace(',', ', ')
        if row.sales_person:
            row.sales_person = row.sales_person.replace(',', ', ')
        
        # Ensure custom_refrence is properly handled
        if not hasattr(row, 'custom_refrence') or row.custom_refrence is None:
            row.custom_refrence = ""
        
        processed_data.append(row)
    
    return processed_data

def get_outstanding_amount(invoice_no):
    """Get outstanding amount for a sales invoice"""
    try:
        return frappe.db.get_value("Sales Invoice", invoice_no, "outstanding_amount") or 0
    except:
        return 0

@frappe.whitelist()
def get_chart_data(filters):
    """Get data for charts, respecting all filters"""
    filters = json.loads(filters) if isinstance(filters, str) else filters
    conditions, condition_values = get_conditions(filters)

    # Daily sales trend
    daily_trend_query = """
        SELECT 
            si.posting_date as date,
            SUM(si.grand_total) as amount,
            SUM(si.discount_amount) as discount
        FROM `tabSales Invoice` si
        WHERE si.docstatus = 1 {conditions}
        GROUP BY si.posting_date
        ORDER BY si.posting_date
    """.format(conditions=conditions)
    
    try:
        daily_data = frappe.db.sql(daily_trend_query, condition_values, as_dict=1)
        
        # Ensure all values are floats for charting
        for item in daily_data:
            item['amount'] = flt(item.get('amount', 0))
            item['discount'] = flt(item.get('discount', 0))
    except Exception as e:
        frappe.log_error(f"Error in daily trend chart: {str(e)}")
        daily_data = []

    # Sales by customer
    customer_sales_query = """
        SELECT 
            si.customer,
            si.customer_name,
            SUM(si.grand_total) as total_sales
        FROM `tabSales Invoice` si
        WHERE si.docstatus = 1 {conditions}
        GROUP BY si.customer, si.customer_name
        ORDER BY total_sales DESC
        LIMIT 10
    """.format(conditions=conditions)
    
    try:
        customer_data = frappe.db.sql(customer_sales_query, condition_values, as_dict=1)
        
        # Ensure all values are floats for charting
        for item in customer_data:
            item['total_sales'] = flt(item.get('total_sales', 0))
    except Exception as e:
        frappe.log_error(f"Error in customer chart: {str(e)}")
        customer_data = []
    
    return {
        "daily_trend": daily_data or [],
        "top_customers": customer_data or []
    }

@frappe.whitelist()
def get_outstanding_summary(filters):
    """Get outstanding amount summary"""
    filters = json.loads(filters) if isinstance(filters, str) else filters
    conditions, condition_values = get_conditions(filters)
    
    # Customer outstanding summary
    customer_outstanding_query = """
        SELECT 
            si.customer,
            si.customer_name,
            SUM(si.grand_total) as total_sales,
            SUM(si.outstanding_amount) as outstanding_amount,
            SUM(si.grand_total - si.outstanding_amount) as total_paid
        FROM `tabSales Invoice` si
        WHERE si.docstatus = 1 {conditions}
        GROUP BY si.customer, si.customer_name
        HAVING outstanding_amount != 0
        ORDER BY outstanding_amount DESC
    """.format(conditions=conditions)
    
    try:
        customer_outstanding = frappe.db.sql(customer_outstanding_query, condition_values, as_dict=1)
        
        # Calculate total outstanding
        total_outstanding = sum(flt(item.get('outstanding_amount', 0)) for item in customer_outstanding)
        
        # Calculate aging summary
        aging_summary = get_aging_summary(filters)
        
    except Exception as e:
        frappe.log_error(f"Error in outstanding summary: {str(e)}")
        customer_outstanding = []
        total_outstanding = 0
        aging_summary = {}
    
    return {
        "customer_outstanding": customer_outstanding,
        "total_outstanding": total_outstanding,
        "aging_summary": aging_summary
    }

def get_aging_summary(filters):
    """Get aging analysis for outstanding amounts"""
    conditions, condition_values = get_conditions(filters)
    
    # Get current date for aging calculation
    current_date = getdate(nowdate())
    
    aging_query = """
        SELECT 
            si.name,
            si.posting_date,
            si.outstanding_amount
        FROM `tabSales Invoice` si
        WHERE si.docstatus = 1 
        AND si.outstanding_amount > 0
        {conditions}
    """.format(conditions=conditions)
    
    try:
        aging_data = frappe.db.sql(aging_query, condition_values, as_dict=1)
        
        range1 = range2 = range3 = range4 = 0  # 0-30, 31-60, 61-90, 90+ days
        
        for invoice in aging_data:
            posting_date = getdate(invoice.get('posting_date'))
            days_old = (current_date - posting_date).days
            outstanding = flt(invoice.get('outstanding_amount', 0))
            
            if days_old <= 30:
                range1 += outstanding
            elif days_old <= 60:
                range2 += outstanding
            elif days_old <= 90:
                range3 += outstanding
            else:
                range4 += outstanding
                
    except Exception as e:
        frappe.log_error(f"Error in aging analysis: {str(e)}")
        range1 = range2 = range3 = range4 = 0
    
    return {
        "range1": range1,
        "range2": range2,
        "range3": range3,
        "range4": range4
    }